RATEST-180: E2E Encounter workflow test

kdaud opened the pull request:
Ticket [ID](https://issues.openmrs.org/browse/RATEST-180)

Description -> Implementation of E2E Encounter workflow test

sherrif10 commented on README.md:
```diff
@@ -14,6 +14,7 @@ ___
   * [![RefApp 2.x Vitals And Triaging Test](https://github.com/openmrs/openmrs-contrib-qaframework/actions/workflows/refapp-2x-vitals-and-triaging.yml/badge.svg)](https://github.com/openmrs/openmrs-contrib-qaframework/actions/workflows/refapp-2x-vitals-and-triaging.yml) 
+  * [![RefApp 2.x Encounter Management Test](https://github.com/openmrs/openmrs-contrib-qaframework/actions/workflows/refapp-2x-encounter.yml/badge.svg)](https://github.com/openmrs/openmrs-contrib-qaframework/actions/workflows/refapp-2x-encounter.yml)
   * Tests in development:
     * Managing User Accounts
```
Do we still need this here on line 19

kdaud replied:
> What does this test basically do

Just as the commit title message states, the test against encounters that the clinician can do with the patient eg `Admission`

sherrif10 replied:
What do you mean by an encounter in openmrs , How do you really understand what an encounter is in openmrs

sherrif10 commented:
This workflow doesnot  seems to be not in line of handling encounters, What are you capturing with in this test, what is an encounter that we are showing the workflow specifically

kdaud commented:
> What are you capturing with in this test,

Refer to the comment [above
](https://github.com/openmrs/openmrs-contrib-qaframework/pull/81#discussion_r692059439)

> what is an encounter that we showing the workflow specifically 

[Admission ](https://github.com/openmrs/openmrs-contrib-qaframework/blob/7566062bd6d66222dd9486db9508c5d154b8f602/src/test/resources/features/refapp-2.x/08-encounter/encounters.feature#L12) [Transfer ](https://github.com/openmrs/openmrs-contrib-qaframework/blob/7566062bd6d66222dd9486db9508c5d154b8f602/src/test/resources/features/refapp-2.x/08-encounter/encounters.feature#L16)and [Discharge](https://github.com/openmrs/openmrs-contrib-qaframework/blob/7566062bd6d66222dd9486db9508c5d154b8f602/src/test/resources/features/refapp-2.x/08-encounter/encounters.feature#L20)

sherrif10 commented:
Left a brand new comment above .Thanks

kdaud commented:
In the world of health, an encounter is an interaction between a patient and healthcare provider(s) for the purpose of providing healthcare service(s) or assessing the health status of a patient.

OpenMRS encounters are  just in line with the explanation above! 
cc: @sherrif10

sherrif10 commented:
> OpenMRS encounters are just in line with the explanation above!

How do we handle an encounter in openmrs then. Your explanation above tells me the only meaning of an encounter.

kdaud commented:
AFAICT the explanation at this [comment ](https://github.com/openmrs/openmrs-contrib-qaframework/pull/81#discussion_r692059439) and as well [here ](https://github.com/openmrs/openmrs-contrib-qaframework/pull/81#issuecomment-901874113) tells well what the test workflow does and its implementation [here](https://github.com/openmrs/openmrs-contrib-qaframework/blob/7566062bd6d66222dd9486db9508c5d154b8f602/src/test/java/org/openmrs/contrib/qaframework/automation/EncountersSteps.java) tells it all.

sherrif10 commented on README.md:
```diff
@@ -14,9 +14,8 @@ ___
   * [![RefApp 2.x Vitals And Triaging Test](https://github.com/openmrs/openmrs-contrib-qaframework/actions/workflows/refapp-2x-vitals-and-triaging.yml/badge.svg)](https://github.com/openmrs/openmrs-contrib-qaframework/actions/workflows/refapp-2x-vitals-and-triaging.yml) 
+  * [![RefApp 2.x Encounter Management Test](https://github.com/openmrs/openmrs-contrib-qaframework/actions/workflows/refapp-2x-encounter.yml/badge.svg)](https://github.com/openmrs/openmrs-contrib-qaframework/actions/workflows/refapp-2x-encounter.yml)
   * Tests in development:
-    * Managing User Accounts
```
Please remove things where applicable , this was supposed to be removed in   https://github.com/openmrs/openmrs-contrib-qaframework/pull/78

kdaud replied:
Did you get a chance to see my comment [here](https://github.com/openmrs/openmrs-contrib-qaframework/pull/78#discussion_r692186811) ?

sherrif10 replied:
Please make things look standard, that commit is supposed to be removed from  RATEST-157

kdaud replied:
See my comment [here](https://github.com/openmrs/openmrs-contrib-qaframework/pull/78#discussion_r692201634)

kdaud commented:
cc: @ibacher

kdaud commented:
Waiting for recommendation cc: @sherrif10

sherrif10 commented on src/test/java/org/openmrs/contrib/qaframework/automation/EncountersSteps.java:
```diff
@@ -0,0 +1,101 @@
+		findPatientPage.enterPatient(testPatient.identifier);
+		findPatientPage.waitForPageToLoad();
+		visitsDashboardPage = findPatientPage.clickOnFirstPatient()
+				.startVisit();
```
Can this be on same line.

sherrif10 replied:
Any reason of inclining who functionalities at the same time.  **Search For exisiting Patient   + starting a visit**

kdaud replied:
After simulating the test when start visit has been made to be start alone from clickOnFirstPatient() function, the test fails when it can not locate startVist() function even when its visible on the clinician dashboard page . Adding `waitForPage()` or `waitForPageToLoad()` does not make the function to be visible either. Investigation indicates that the issues are originating from the way how the page was designed in [distro](https://github.com/openmrs/openmrs-distro-referenceapplication/tree/master/ui-tests/src/test/java/org/openmrs/reference/page) and making a refactoring to suit the requirement will fail many selenium legacy test by default. 

Yes start Visit can be visible for the test to pass when the test logic has been refactored to pass via the Active visit but this wing does not prevail the encounters that are under test and makes the suggested wing to override the alternative.

kdaud replied:
> Can this be on the same line 

We are glad that BDD framework workflow builds `auto-formats` the code base when they are being run locally. It would look very nice if we had some thing like `visitsDashboardPage.goToTransferToWardServicePage().confirm("Isolation Ward");` on the same line than having two lines at `73` and `74` but this is as result of auto formatting  that is than during the build plan.

Putting  it on the same line would make our code base look well I would say however, contributors working on new workflows will find a challenge when committing their changes since during the local build plan the workflow will auto-format all the existing test files and they will by default be tracked as changed lines introduced and this might not be a good move to trend.

When you look at other existing definition steps have this kind of format though it does not look good to me as well ! Example [FindPatientSteps](https://github.com/openmrs/openmrs-contrib-qaframework/blob/6a612c2ea4a56af898074d3227f026eff8010749/src/test/java/org/openmrs/contrib/qaframework/automation/FindPatientSteps.java#L29) [ConditionsSteps](https://github.com/openmrs/openmrs-contrib-qaframework/blob/6a612c2ea4a56af898074d3227f026eff8010749/src/test/java/org/openmrs/contrib/qaframework/automation/ConditionsSteps.java#L36) [ClinicalVisitSteps](https://github.com/openmrs/openmrs-contrib-qaframework/blob/6a612c2ea4a56af898074d3227f026eff8010749/src/test/java/org/openmrs/contrib/qaframework/automation/ClinicalVisitSteps.java#L81)

sherrif10 replied:
> the test fails when it can not locate startVist() function even when its visible on the clinician dashboard page

This is amazing work  .Why is the test failing to start a visit when other start visit functionalities have been working perfect, Am honestly not understanding this workflow. Do you mind sharing us where it fails we add a debugger point please.

sherrif10 replied:
> but this wing does not prevail the encounters that are under test and makes the suggested wing to override the alternative.

What do you mean by wing

kdaud replied:
> the test fails when it can not locate startVist() function even when its visible on the clinician dashboard page
> Why is the test failing to start a visit when other start visit functionalities have been working perfect

After fixing the issue which was originating from the server, the test logic as suggested by @sherrif10 worked liked  a charm !

sherrif10 commented on src/test/java/org/openmrs/contrib/qaframework/automation/EncountersSteps.java:
```diff
@@ -0,0 +1,101 @@
+	@When("a user clicks on Trasfer to Ward service button and selects location")
+	public void goToTrasferToWardServicePage() {
+		visitsDashboardPage.goToTransferToWardServicePage().confirm(
+				"Isolation Ward");
```
Can this be in one line as well.

kdaud replied:
Refer to the comment [above](https://github.com/openmrs/openmrs-contrib-qaframework/pull/81#discussion_r693478959)

sherrif10 commented on src/test/java/org/openmrs/contrib/qaframework/automation/EncountersSteps.java:
```diff
@@ -0,0 +1,101 @@
+	}
+
+	@Then("the system loads patient visit dashboard page")
+	public void systemloadsVisitDashboardPage() {
```
Is it visitDashboardPage or visitPage.

kdaud replied:
The method loads patient visit dashboard page

sherrif10 replied:
> The method loads patient visit dashboard page

Very true, thanks for clarification , have confirmed it[ here]([url](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/master/ui-tests/src/test/java/org/openmrs/reference/page/PatientVisitsDashboardPage.java)). how about calling it exact name , **LoadPatientVisitDashboardPage()**

sherrif10 commented on src/test/java/org/openmrs/contrib/qaframework/automation/EncountersSteps.java:
```diff
@@ -0,0 +1,101 @@
+		assertTrue(visitsDashboardPage.containsText("Discharge"));
+		assertThat(visitsDashboardPage.getEncountersCount(), is(3));
+		visitsDashboardPage.waitForPageToLoad();
+		goToPatientDashboard();
```
Can we make this function as a separate step, we might even use it in future in case this tests get too much traffice

kdaud replied:
I thought about the same idea but i realized its not align with the use story for encounter. Actually the user story should end when the 3 encounters have been confirmed at  line line [22](https://github.com/openmrs/openmrs-contrib-qaframework/blob/57bb3fc5467059aa75fb19303cd1e3e5aabf81a5/src/test/resources/features/refapp-2.x/08-encounter/encounters.feature#L22)

From the technical end I thought it very important for the test to return to the clinician dashbaord before it quits the run so that these encounters are observed on the dashbaord. Looking into how the idea would be embedded into the feature file makes it odd man out since its not a user story in the context and I decided to just create a method called [goToPatientDashboard()](https://github.com/openmrs/openmrs-contrib-qaframework/blob/57bb3fc5467059aa75fb19303cd1e3e5aabf81a5/src/test/java/org/openmrs/contrib/qaframework/automation/EncountersSteps.java#L97) and invoke it in the last user story [here](https://github.com/openmrs/openmrs-contrib-qaframework/blob/57bb3fc5467059aa75fb19303cd1e3e5aabf81a5/src/test/java/org/openmrs/contrib/qaframework/automation/EncountersSteps.java#L94)

sherrif10 replied:
> I decided to just create a method called [goToPatientDashboard()](https://github.com/openmrs/openmrs-contrib-qaframework/blob/57bb3fc5467059aa75fb19303cd1e3e5aabf81a5/src/test/java/org/openmrs/contrib/qaframework/automation/EncountersSteps.java#L97) and invoke it in the last user story [here](https://github.com/openmrs/openmrs-contrib-qaframework/blob/57bb3fc5467059aa75fb19303cd1e3e5aabf81a5/src/test/java/org/openmrs/contrib/qaframework/automation/EncountersSteps.java#L94)

I think for better re-using of these methods, we need to extract them , i think you better write another method that handles goToPatientDashboard. this will help us in future to re use this method. this is the same problem that is occuring in [transfertest](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/f2e3df96f7ff815ee8249c4dce61ae460339f770/ui-tests/src/test/java/org/openmrs/reference/page/TransferPatientPage.java#L23) , it will save us time.Thanks

kdaud replied:
> I think for better re-using of these methods, we need to extract them , i think you better write another method that handles goToPatientDashboard. this will help us in future to re use this method

I agree!

To make the method re-susable, I actually created this method earlier on [here](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/f2e3df96f7ff815ee8249c4dce61ae460339f770/ui-tests/src/test/java/org/openmrs/reference/page/PatientVisitsDashboardPage.java#L99) and its just being invoked at code line [100](https://github.com/openmrs/openmrs-contrib-qaframework/blob/d0f633b4f463ba72b6b27d017f738f8513e0aea9/src/test/java/org/openmrs/contrib/qaframework/automation/EncountersSteps.java#L100)

kdaud commented on src/test/java/org/openmrs/contrib/qaframework/automation/EncountersSteps.java:
```diff
@@ -0,0 +1,101 @@
+	// Admission encounter
+	@When("a user clicks on Admit to Inpatient button and selects location")
+	public void goToAdmitToInpatientPage() {
+		visitsDashboardPage.goToAdmitToInpatient().confirm("Inpatient Ward");
```
It would have been nice to have them separated but from the design of the page, the method confirm can only be accessed using the method goToAdmitToInpatient() other than that the method can not be accessible. I found out the issue are originating from the design of the page as observed [here](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/f2e3df96f7ff815ee8249c4dce61ae460339f770/ui-tests/src/test/java/org/openmrs/reference/page/TransferPatientPage.java#L23).

Refactoring the page will by default fail the selenium tests that consume the function and personally would not prefer taking time to dive into refactoring the page but rather focus on writing tests for functionalities that were not covered by the selenium test.

Yes if we are to look into refactoring the[ page](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/f2e3df96f7ff815ee8249c4dce61ae460339f770/ui-tests/src/test/java/org/openmrs/reference/page/TransferPatientPage.java#L6) then we can create a ticket to cater for  that and would be good for new contributors to kick off with such tickets But this should not put this test in a waiting state!

sherrif10 replied:
> Yes if we are to look into refactoring the[ page](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/f2e3df96f7ff815ee8249c4dce61ae460339f770/ui-tests/src/test/java/org/openmrs/reference/page/TransferPatientPage.java#L6) then we can create a ticket to cater for that and would be good for new contributors to kick off with such tickets But this should not put this test in a waiting state!

This is fine we can leave this for now

kdaud commented:
Waiting for recommendation cc: @sherrif10

kdaud commented:
Waiting for recommendation cc: @sherrif10

kdaud commented:
Waiting for recommendation cc: @sherrif10

sherrif10 commented:
Thanks @kdaud , Have added some few comments on the ticket

kdaud commented:
Thanks @sherrif10 for reviews. Have addressed them and waiting for recommendation

kdaud commented:
Waiting for final review to see this workflow test running in master branch cc: @dkayiwa

sherrif10 commented:
Will test this ,cc @dkayiwa, I still believe there is some work to be done.

On Mon, Aug 23, 2021, 1:10 PM Kakumirizi Daud ***@***.***>
wrote:

> Waiting for final review to see this workflow test running in master
> branch cc: @dkayiwa <https://github.com/dkayiwa>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openmrs/openmrs-contrib-qaframework/pull/81#issuecomment-903629755>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALOKS6COCLJES6ZQ63NK6DTT6INAZANCNFSM5CN436BA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>
> .
>

dkayiwa commented:
@sherrif10 that is why i will not merge before your approval. 😊

kdaud commented:
> I still believe there is some work to be done.

Am eager to work on it if am given a hint on which task exactly is still needed otherwise have tested the workflow on my local instance with all the browser instances and its passing well. GithubActions has already confirmed the tests and Bamboo will finally affirm
cc: @sherrif10  @dkayiwa

kdaud commented:
Waiting for recommendation cc: @sherrif10

sherrif10 commented:
Hey @kdaud , sorry for a delaying a bit however am seeing encounters at a big perspective, Am still imagining which exact steps can be added in encounters, FYI encounters in openmrs usually have one providers and in many cases there is morethan one provider in an encounter.

kdaud commented:
> Am still imagining which exact steps can be added in encounters

A mentioned on slack the beauty with the BDD framework we can always scale the workflow any time (I mean adding user stories in an existing feature file and writing its corresponding steps definition). For the case of `Vitals` and `Visit Note`  these were covered by clinical visit workflow test and including them as part of encounters (if they qualify) would be ok But this will be duplicating steps which Cucumber does not support and by default the workflow will fail the run and as well imply duplicating efforts on functionality that are already being tested. (edited) 
cc: @sherrif10

sherrif10 commented:
> Cucumber does not support

Cucumber doesnot support duplicate of steps but there is a way to handle that using inheritance.or DI

sherrif10 commented on src/test/java/org/openmrs/contrib/qaframework/automation/InpatienttSteps.java:
```diff
@@ -0,0 +1,103 @@
+	// Discharge encounter
+	@When("a user clicks on Exit from Inpatient button and selects location")
+	public void goToExitFromInpatientPage() {
+		visitsDashboardPage.goToExitFromInpatient().confirm("Isolation Ward");
```
Can we use a different location here may be Inpatient Ward and get this merged.

kdaud replied:
According to the test workflow, I mimicked a real world situation where a patient is admitted in [Inpatient-Ward](https://github.com/openmrs/openmrs-contrib-qaframework/blob/534d5e28cffa213f003ca12a625eff0795d47958/src/test/java/org/openmrs/contrib/qaframework/automation/InpatienttSteps.java#L63) then is transferred to [Isolation-Ward](https://github.com/openmrs/openmrs-contrib-qaframework/blob/534d5e28cffa213f003ca12a625eff0795d47958/src/test/java/org/openmrs/contrib/qaframework/automation/InpatienttSteps.java#L75) then is discharged from [Isolation-Ward](https://github.com/openmrs/openmrs-contrib-qaframework/blob/534d5e28cffa213f003ca12a625eff0795d47958/src/test/java/org/openmrs/contrib/qaframework/automation/InpatienttSteps.java#L88). By changing the location to Inpatient Ward at code line 88 would mean changing other correspoding Wards to bring out a real world user story and would proposed to go which the one suggested in the commit.

sherrif10 commented:
This looks very good @kdaud , Well done, Have left one comment. Then we can approve and merge this.

kdaud commented:
> Have left one comment

Have addressed the comment [here](https://github.com/openmrs/openmrs-contrib-qaframework/pull/81#discussion_r695361524) 

Hey @sherrif10 I would suggest you either recommend or approve, then we wait for the geeks for their final reviews.

sherrif10 reviewed (approved):
Thanks @kdaud   .  cc @dkayiwa  LGTM , its ready to be merged

kdaud commented:
Apart from a few comments if not none that @dkayiwa may give concerning the PR, I guess when he lands on this commit his hands will trigger a merge action to welcome the workflow test embedded in the commit into the world of master builds!

dkayiwa commented on src/test/java/org/openmrs/contrib/qaframework/automation/InpatienttSteps.java:
```diff
@@ -0,0 +1,103 @@
+		assertTrue(visitsDashboardPage.containsText("Visits"));
+	}
+
+	// Admission encounter
```
What value does the above comment add?

kdaud replied:
> What value does the above comment add?

Very interesting question !

Well it has no effect when building the code base. However, I thought it would be of value to include some important comments for  current and future technical devs to know where each user story is actually implemented. As we have three user stories for encounters ie [Admission](https://github.com/openmrs/openmrs-contrib-qaframework/blob/534d5e28cffa213f003ca12a625eff0795d47958/src/test/resources/features/refapp-2.x/08-encounter/inpatient.feature#L13), [Transfer](https://github.com/openmrs/openmrs-contrib-qaframework/blob/534d5e28cffa213f003ca12a625eff0795d47958/src/test/resources/features/refapp-2.x/08-encounter/inpatient.feature#L17) and [Discharge](https://github.com/openmrs/openmrs-contrib-qaframework/blob/534d5e28cffa213f003ca12a625eff0795d47958/src/test/resources/features/refapp-2.x/08-encounter/inpatient.feature#L21) I felt compelled to do add comment in the selenium implementation so that in case any of the user story is failing in future it lessen the geeks time to figure out its equivalent selenium implementation.

dkayiwa replied:
Can't you do that in the method name itself?

dkayiwa replied:
Also, what do you think about this? https://improvingsoftware.com/2011/06/27/5-best-practices-for-commenting-your-code/

kdaud replied:
> Can't you do that in the method name itself?

This is more than an opinion but a great idea. Am now a consultant to myself and qualify to guide others in this context.

> Also, what do you think about this? https://improvingsoftware.com/2011/06/27/5-best-practices-for-commenting-your-code/

Have kept the resource in my archive for future reference.  Its precise and brief with a well defined content embedded!

dkayiwa commented on src/test/java/org/openmrs/contrib/qaframework/automation/InpatienttSteps.java:
```diff
@@ -0,0 +1,103 @@
+		assertThat(visitsDashboardPage.getEncountersCount(), is(1));
+	}
+
+	// Transfer encounter
```
What value does the above comment add?

kdaud replied:
Refer to my comment [above](https://github.com/openmrs/openmrs-contrib-qaframework/pull/81#discussion_r695640449)

dkayiwa commented on src/test/java/org/openmrs/contrib/qaframework/automation/InpatienttSteps.java:
```diff
@@ -0,0 +1,103 @@
+		assertThat(visitsDashboardPage.getEncountersCount(), is(2));
+	}
+
+	// Discharge encounter
```
What value does the above comment add?

kdaud replied:
Refer to my comment [above](https://github.com/openmrs/openmrs-contrib-qaframework/pull/81#discussion_r695640449)

sherrif10 commented on src/test/java/org/openmrs/contrib/qaframework/automation/InpatienttSteps.java:
```diff
@@ -57,9 +57,8 @@ public void systemloadsPatientVisitDashboardPage() {
 
-	// Admission encounter
 	@When("a user clicks on Admit to Inpatient button and selects location")
-	public void goToAdmitToInpatientPage() {
```
Seems you are messing up with things here. This is the right way of naming the method

kdaud replied:
> Seems you are messing up with things here

Have you had a glimpse at @dkayiwa 's comment [here](https://github.com/openmrs/openmrs-contrib-qaframework/pull/81#discussion_r695641883) ? gotToInpatientPage() is accomplishing Admission encounter in the nutshell and makes meaning to me from the technical angle to rename it to `admissionEncounter()` so as to eliminate un necessary comment(s) within our code base.

kdaud commented:
All is well that ends well !! 
cc: @dkayiwa

dkayiwa commented on src/test/java/org/openmrs/contrib/qaframework/automation/InpatienttSteps.java:
```diff
@@ -0,0 +1,100 @@
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.junit.Assert.assertTrue;
+
+public class InpatienttSteps extends Steps {
```
Any extra `t`?

dkayiwa commented on src/test/java/org/openmrs/contrib/qaframework/automation/InpatienttSteps.java:
```diff
@@ -0,0 +1,100 @@
+	}
+
+	@When("a user clicks on Admit to Inpatient button and selects location")
+	public void admissionEncounter() {
```
The new naming convention looks different from what you had before and else where. The original name explicitly specified what was to be done. For instance, `clickOnStartVisit, searchExistingPatient,  goToFindPatientRecordPage `. I find the older convention clearer and more explicit about what is to be done than `admissionEncounter, transferEncounter , dischargeEncounter `

kdaud replied:
The commit with id `317b2b1` explains the response to the review  [above](https://github.com/openmrs/openmrs-contrib-qaframework/pull/81#discussion_r695753689).

kdaud commented:
Keeping my eyes open to seeing the event mentioned [here](https://github.com/openmrs/openmrs-contrib-qaframework/pull/81#issuecomment-905374283) to happen. 
cc: @dkayiwa

dkayiwa commented on src/test/java/org/openmrs/contrib/qaframework/automation/InpatientSteps.java:
```diff
@@ -0,0 +1,100 @@
+	}
+
+	@Then("the system loads patient visit dashboard page")
+	public void systemloadsPatientVisitDashboardPage() {
```
systemloads or systemLoads?

dkayiwa commented:
And the merge conflicts.
